corehq/apps/userreports/tests/test_choice_provider.py

```diff
@@ -0,0 +1,305 @@
+from abc import ABCMeta, abstractmethod
+from django.test import SimpleTestCase, TestCase
+import mock
+from corehq.apps.commtrack.tests.util import bootstrap_location_types, make_loc
+from corehq.apps.es.fake.groups_fake import GroupESFake
+from corehq.apps.es.fake.users_fake import UserESFake
+from corehq.apps.groups.models import Group
+from corehq.apps.locations.tests import delete_all_locations
+from corehq.apps.reports_core.filters import Choice
+from corehq.apps.userreports.models import ReportConfiguration
+from corehq.apps.userreports.reports.filters.choice_providers import ChoiceProvider, \
+    ChoiceQueryContext, LocationChoiceProvider, UserChoiceProvider, GroupChoiceProvider, \
+    OwnerChoiceProvider
+from corehq.apps.users.models import CommCareUser, WebUser, DomainMembership
+from corehq.apps.users.util import normalize_username
+
+
+class SearchableChoice(Choice):
+    def __new__(cls, value, display, searchable_text=None):
+        self = super(SearchableChoice, cls).__new__(cls, value, display)
+        self.searchable_text = [text for text in searchable_text or []
+                                if text is not None]
+        return self
+
+
+class StaticChoiceProvider(ChoiceProvider):
+
+    def __init__(self, choices):
+        """
+        choices must be passed in in desired sort order
+        """
+        self.choices = [
+            choice if isinstance(choice, SearchableChoice)
+            else SearchableChoice(
+                choice.value, choice.display,
+                searchable_text=[choice.display]
+            )
+            for choice in choices
+        ]
+        super(StaticChoiceProvider, self).__init__(None, None)
+
+    def query(self, query_context):
+        filtered_set = [choice for choice in self.choices
+                        if any(query_context.query in text for text in choice.searchable_text)]
+        return filtered_set[query_context.offset:query_context.offset + query_context.limit]
+
+    def get_choices_for_known_values(self, values):
+        return {choice for choice in self.choices
+                if choice.value in values}
+
+
+class StaticChoiceProviderTest(SimpleTestCase):
+    @classmethod
+    def setUpClass(cls):
+        cls.choice_provider = StaticChoiceProvider([Choice('1', 'One'), Choice('2', 'Two'), Choice('3', 'Three')])
+
+    def test_query_no_search(self):
+        self.assertEqual(self.choice_provider.query(ChoiceQueryContext('', 2, page=0)),
+                         [Choice('1', 'One'), Choice('2', 'Two')])
+        self.assertEqual(self.choice_provider.query(ChoiceQueryContext('', 2, page=1)),
+                         [Choice('3', 'Three')])
+
+    def test_get_choices_for_values(self):
+        self.assertEqual(
+            set(self.choice_provider.get_choices_for_values(['2', '4', '6'])),
+            {Choice('2', 'Two'), Choice('4', '4'), Choice('6', '6')}
+        )
+
+
+class ChoiceProviderTestMixin(object):
```

Comment:
took me a bit to figure out how this worked. might benefit from a small doc string explaining how you should setup one real and one static choice provider that you expect to behave the same and then use this to compare them